From 56c68cbdfcdc0c8359790d8bdf2cdfbb0684eeca Mon Sep 17 00:00:00 2001
From: Marc Mao <marc.mao@intel.com>
Date: Fri, 16 Jun 2023 16:24:35 +0530
Subject: [PATCH] Add external buffer tiling flag for tiled format

TODO: to support tiling format better, we should use modifier
information with PRIME2 API.

Change-Id: I39a79b37bfa5658230ebb42846dc0651fba31aa4
Tracked-On: OAM-110710
Signed-off-by: Marc Mao <marc.mao@intel.com>
---
 c2_utils/src/mfx_va_allocator.cpp | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/c2_utils/src/mfx_va_allocator.cpp b/c2_utils/src/mfx_va_allocator.cpp
index 7047955..741b38d 100755
--- a/c2_utils/src/mfx_va_allocator.cpp
+++ b/c2_utils/src/mfx_va_allocator.cpp
@@ -111,6 +111,19 @@ static mfxU32 ConvertVAFourccToVARTFormat(mfxU32 va_fourcc)
     }
 }
 
+// quick fix tiling issue, formal should use modifier with PRIME2 interface
+static bool isTiledGrallocFourcc(int fourcc) {
+    switch (fourcc) {
+        case HAL_PIXEL_FORMAT_NV12_Y_TILED_INTEL:
+        case HAL_PIXEL_FORMAT_YUV420PackedSemiPlanar_Tiled_INTEL:
+        case HAL_PIXEL_FORMAT_P010_INTEL:
+            return true;
+        default:
+            break;
+    }
+    return false;
+}
+
 MfxVaFrameAllocator::MfxVaFrameAllocator(VADisplay dpy)
     : m_dpy(dpy)
 {
@@ -546,6 +559,8 @@ mfxStatus MfxVaFrameAllocator::CreateSurfaceFromGralloc(const MfxGrallocModule::
     if (IS_PRIME_VALID(info.prime)) {
         surfExtBuf.buffers = (uintptr_t *)&info.prime;
         surfExtBuf.flags = VA_SURFACE_ATTRIB_MEM_TYPE_DRM_PRIME;
+        if (isTiledGrallocFourcc(info.format))
+            surfExtBuf.flags |= VA_SURFACE_EXTBUF_DESC_ENABLE_TILING;
     } else {
         surfExtBuf.buffers = (uintptr_t *)&info.handle;
         surfExtBuf.flags = VA_SURFACE_ATTRIB_MEM_TYPE_ANDROID_GRALLOC;
-- 
2.40.1

